compiler/src/iree/compiler/Preprocessing/Common/ConvertConvFilterToChannelsLast.cpp

```diff
@@ -0,0 +1,144 @@
+//===----------------------------------------------------------------------===//
+// ConvertConvFilterToChannelsLastPass
+//===----------------------------------------------------------------------===//
+
+#include "iree/compiler/Preprocessing/Common/Passes.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/LogicalResult.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-preprocessing-convert-conv-filter-to-channels-last"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+namespace mlir::iree_compiler::Preprocessing {
+
+#define GEN_PASS_DEF_CONVERTCONVFILTERTOCHANNELSLASTPASS
+#include "iree/compiler/Preprocessing/Common/Passes.h.inc"
+
+// Utility function to swap the last two dimensions.
+static AffineMap constructFilterMap(AffineMap map, SmallVector<int64_t> &perm) {
+  unsigned numDims = map.getNumDims();
+  ArrayRef<AffineExpr> mapResults = map.getResults();
+  SmallVector<AffineExpr, 4> exprs;
+  for (int i = 0; i < perm.size(); ++i) {
+    exprs.push_back(mapResults[perm[i]]);
+  }
+  return AffineMap::get(numDims, map.getNumSymbols(), exprs, map.getContext());
+}
+
+// Utility function to create a transpose operation.
+static Value createTransposeOp(PatternRewriter &rewriter, Location loc,
+                               Value tensor, SmallVector<int64_t> &perm) {
+  SmallVector<OpFoldResult, 4> dimSizes =
+      tensor::getMixedSizes(rewriter, loc, tensor);
+  // Dim index of H, W, F, C
+  SmallVector<OpFoldResult, 4> transposeResultDims;
+  for (int i = 0; i < perm.size(); ++i) {
+    transposeResultDims.push_back(dimSizes[perm[i]]);
+  }
+
+  auto tensorType = cast<RankedTensorType>(tensor.getType());
+  auto emptyTensor = rewriter.create<tensor::EmptyOp>(
+      loc, transposeResultDims, tensorType.getElementType());
+  return rewriter.create<linalg::TransposeOp>(loc, tensor, emptyTensor, perm)
+      .getResult()[0];
+}
+
+llvm::LogicalResult
+convertConvFilterToTargetLayout(linalg::Conv2DNhwcHwcfOp convOp,
+                                PatternRewriter &rewriter,
+                                SmallVector<int64_t> &perm) {
+  Location loc = convOp.getLoc();
+
+  // Extract operands.
+  Value input = convOp.getInputs()[0];
+  Value filter = convOp.getInputs()[1];
+  Value output = convOp.getOutputs()[0];
+
+  // Extract indexing maps.
+  AffineMap inputMap = convOp.getIndexingMapsArray()[0];
+  AffineMap filterMap = convOp.getIndexingMapsArray()[1];
+  AffineMap outputMap = convOp.getIndexingMapsArray()[2];
+
+  AffineMap transposedFilterMap = constructFilterMap(filterMap, perm);
```

Comment:
Can this be done with `AffineMap::getPermutationMap` + `compose`? E.g.,

```
auto transposeMap = AffineMap::getPermutationMap::get(perm, ctx);
AffineMap transposedFilterMap = filterMap.compose(tranposeMap);
```

Reply:
Just gave it a shot, originally, 
filterMap is: `(d0, d1, d2, d3, d4, d5, d6) -> (d4, d5, d6, d3)`
transposedFilterMap is: `(d0, d1, d2, d3, d4, d5, d6) -> (d4, d5, d3, d6)`

With the suggested code:
transposeMap is: `(d0, d1, d2, d3) -> (d0, d1, d3, d2)`
With `filterMap.compose`, it will attempt compose filterMap with transposeMap. But then since filterMap's dimension is 7 and transposeMap's results is 4, it'd run into the below assertion:

https://github.com/iree-org/llvm-project/blob/6a722c20b272dac9a0e5a2a27a06587b15adb0a3/mlir/lib/IR/AffineMap.cpp#L557

I think `afineMap::compose` was intentionally being conservative and exclude the situation where dimensions are dropped from results.

Nice suggestion though! Glad I learned a bit more about infrastructure of AffineMap.

Reply:
Oh, I missed that the indexing map of the filter map is not a permutation map. Sorry for the bad suggestion, and thanks that you're okay with it. :)

Reply:
There's no bad suggestions :-p. Thank you very much with your detailed code review.